presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java

```diff
@@ -419,11 +419,38 @@ protected GroupedExecutionProperties visitPlan(PlanNode node, Void context)
         @Override
         public GroupedExecutionProperties visitJoin(JoinNode node, Void context)
         {
-            GroupedExecutionProperties properties = processChildren(node);
-            if (properties.isCurrentNodeCapable()) {
-                return new GroupedExecutionProperties(true, true);
+            GroupedExecutionProperties left = node.getLeft().accept(this, null);
+            GroupedExecutionProperties right = node.getRight().accept(this, null);
+
+            if (Optional.empty().equals(node.getDistributionType())) {
+                // This is possible when the optimizers is invoked with `forceSingleNode` set to true.
```

Comment:
I think we should be able to pull `DetermineJoinDistributionType` and `DetermineSemiJoinDistributionType` outside of `if (!forceSingleNode)` check in the `PlanOptimizers`

Reply:
Feel free to put up a PR for that.